### ISSUE TITLE
A little more clarity please, will help transfer api from Craft 2

### ISSUE THREAD

**Issue by narration-sd** [craftcms/element-api#52]
I had a usual session of tracing though what Craft was actually doing with its abstractions last night, to get this plugin running, which it actually does.

You could make this all much easier and accessible by making the following clear:

- the name of the config file now has to be element-api.php, not elementapi.php as before. No doubt due  to the snake-case worldview now present, but mis-stated in your plugin doc. In which case, all you get is 404s.

- when stating that, very clarifying would be to say that element-api.php is handled by the standard Yii-Craft config system; thus all related patterns apply, and "endpoints" are just an array of routing paths with their dispositions

- thus, what you put in your given endpoint hard-defines the path that will access its element(s).

- so, if you want //domain/api/endpoint, specify "api/endpoint.json" as the name. If you want simply //domain/endpoint to be the access, use "/endpoint.json" instead.

- your example illustrates only the second case, which differs from how Craft 2's elementapi showed it, hence more unnecessary 404s.  I'd think including the uri for the api call would be a very good idea, not to say mentioning you're showing a fresh style, adjustable to preference.

**Comment by brandonkelly** [comment]
> your example illustrates only the second case, which differs from how Craft 2's elementapi showed it, hence more unnecessary 404s. I'd think including the uri for the api call would be a very good idea, not to say mentioning you're showing a fresh style, adjustable to preference.

Not sure I’m following you here - what changed in the v2 docs vs. v1?

**Comment by narration-sd** [comment]
Hmm. Changelog shows you fixed the elementapi.php element-api.php change yesterday, looked as I was surprised and thought I'd slipped over it originally, but no.

As far as the naming of endpoints, correct must say, nothing has changed that I could easilly find, as they stand now. Though I could think to remember that  api/endpoint figured in there earlier days when I did a v1 coding. Maybe I just invented it, but think not?? Haven't parsed all your plugin commits to see.

My overall point is this. Scanning the doc again, you've suggested learning-by-example(s). I generally think this is a good plan for picking up detail points; it's at least making a compression of writing (and thus reading) necessary to cover all that's possible with what you made. So ok.

But. While I think example-detail-showing can be prioritized, I also feel it's pretty necessary to have the basis concepts of your abstractions (i.e. 'api' is the abstraction for the plugin here) outlined clearly as a first basis.

If you don't do that, we have to discover it, any time things don't go. As in, getting 404s twice over due to the element-api introducer change, and then because of needing/not-needing api/ to preface endpoints. The two cases multiplly, of course, and 404s are not visibly clues, if they actually are.

Discovery means in many cases not just reading code, but tracing through operations via PhpStorm or other. I admit to learn a great deal this way, but Craft has an overhead due to its multifaceted design. I takes me time, also effort to decide what (other abstractions) to accept/ignore, so I can get to the point.  

For others, esp. 'designers' in the webcode sense and who would use an api, I'm not confident in the chops or patience.

The tracing is to grasp the _meaning_ of your words or codes, those 'abstractions'.  

Ok, an api. Ok, an endpoint. But to discover that an 'endpoint' is exactly a route, makes the eyes truly open. 

Now the understanding can be there, for the examples, and for inventing on your own.

'Routes', btw, as far as i remember, would be another of those slightly cloudy concepts, also requiring about a one-liner to make clear. You see my principle...am sure, Brandon.

Words, meanings, kind of my own more base specialities, I guess, yes. And yes, I was a little cranky when writing in yesterday about this, after the episode; comes to us all. I do try to watch the poetics that want to creep in, my own way of confusing, besides tantilizing away boredom :) .

Take care, my friend.
